pkg/operator/ceph/cluster/cluster.go

```diff
@@ -138,16 +138,17 @@ func (c *cluster) detectCephVersion(image string, timeout time.Duration) (*cephv
 }
 
 func (c *cluster) createInstance(rookImage string, cephVersion cephver.CephVersion) error {
-	if c.checkSetOrchestrationRunning() == true {
-		logger.Debugf("As createInstance is currently running added this request as pending.")
+	// do not start an orchestration run if there is currently one running
+	if c.getOrchestrationRunning() == true {
+		// mark this request as pending
 		c.setOrchestrationPending()
 		return nil
 	}
 
-	defer c.unsetOrchestrationRunning()
-	initRun := true
-	for c.checkUnsetOrchestrationPending() == true || initRun == true {
-		initRun = false
+	defer c.unsetOrchestrationStatus()
```

Comment:
I don't think these actions are atomic enough. There is a race condition possible still (though very unlikely) where the for loop checks `c.checkSetOrchestrationStatus()`, which returns false, at which point the scheduler pauses this routine. About this time, another call to `createInstance` is made, and `c.getOrchestrationRunning` returns true, causing `c.setOrchestrationPending()` to be called. The scheduler continues with the paused routine, which returns nil and executes the `defer`, setting `orchestrationRunning = false` and `orchestrationPending = true`. But now the operator *should* be updating the cluster, but it won't until another update is made (or the operator is restarted), resulting in a new call to `createInstance`.

I think the for loop logic needs to be totally separated from the `createInstance` method. What about this flow ... ?
1. Start a go routine that's purpose is to check if `orchestrationPending == true`, and if so, set `orchestrationRunning = true`, and then call the original `createInstance`. Once `createInstance` returns, `orchestrationRunning` should be set back to `false`. If another orchestration has been made `pending` during the orchestration run, the loop will see that on the next iteration.
2. Instead of calling `createInstance` from the detection of a cluster change event, simply call a method to set `orchestrationPending = true` and return. This may have to update a stored version of the latest pending `rookImage` and `cephVersion` for item 1 above.
3. On operator startup, `orchestrationPending` should still be set to `true`.

I think with this flow, there will be no misleading variable names, and the possible race condition will be eliminated.

Reply:
I had also the idea with a separate routine for `createInstance`. But while checking the code I had to notice that there are subsequent calls which rely on the results of `createInstance`.
This means that this solution is not easily possible.
Therefore I moved all get and set calls of the orchestration variables to `c.checkSetOrchestrationStatus()`. No race conditions should be possible anymore.